File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/DomusEngineRest/GetEndpoint.java
package it.achdjian.paolo.domusviewer.DomusEngineRest;

import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;

import it.achdjian.paolo.domusviewer.zigbee.ZEndpoint;

/**
 * Created by Paolo Achdjian on 18/04/16.
 */
public class GetEndpoint extends DomusEngineRest {
    private static ObjectMapper MAPPER = new ObjectMapper();

    public interface Listener {
        void newEndpoint(ZEndpoint zEndpoint);
    }

    private final Listener listener;
    private final int nwkAddress;
    private final int endpointId;

    public GetEndpoint(@NonNull SharedPreferences sharedPreferences, @NonNull ConnectionStatus connected, int nwkAddress, int endpointId, @NonNull Listener listener) {
        super(sharedPreferences, connected);
        this.nwkAddress = nwkAddress;
        this.endpointId = endpointId;
        this.listener = listener;
    }

    @Override
    public void run() {
        String body = get("/devices/" + Integer.toString(nwkAddress, 16) + "/endpoint/" + Integer.toString(endpointId, 16));
        if (body != null) {
            if (!body.isEmpty()) {
                try {
                    ZEndpoint zEndpoint = MAPPER.readValue(body, ZEndpoint.class);
                    listener.newEndpoint(zEndpoint);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: ZTopology/app/src/main/java/it/achdjian/paolo/ztopology/rest/GetEndpoint.kt
package it.achdjian.paolo.ztopology.domusEngine.rest

import android.util.Log
import it.achdjian.paolo.ztopology.zigbee.ZEndpoint
import it.achdjian.paolo.ztopology.zigbee.ZEndpointJSon
import it.achdjian.paolo.ztopology.DomusEngine
import it.achdjian.paolo.ztopology.MessageType
import it.achdjian.paolo.ztopology.domusEngine.rest.DomusEngineRest
import it.achdjian.paolo.ztopology.domusEngine.rest.ZigbeeRunnable
import java.io.IOException

/**
 * Created by Paolo Achdjian on 7/11/17.
 */
class GetEndpoint(val nwkAddress: Int, val endpoint_id: Int) : ZigbeeRunnable() {

    override fun run() {
        val body = DomusEngineRest.get("/devices/" +nwkAddress.toString(16)+"/endpoint/" + endpoint_id.toString(16))
        if (body.isNotBlank()) {
            try {
                Log.i("REST", body)
                val endpointJson = MAPPER.readValue(body, ZEndpointJSon::class.java)
                if (endpointJson != null) {
                    val zEndpoint = ZEndpoint(endpointJson)
                    Log.i("REST",zEndpoint.toString())
                    DomusEngine.handler.sendMessage(DomusEngine.handler.obtainMessage(MessageType.NEW_ENDPOINT, zEndpoint))
                }
            } catch (e: IOException) {
                Log.e(TAG, "Error parsing response for /devices/$nwkAddress/endpoint/$endpoint_id", e)
                Log.e(TAG, "Response: " + body)
                e.printStackTrace()
            }
        }
    }
}